module/zfs/vdev_raidz.c

```diff
@@ -2299,15 +2344,109 @@ vdev_raidz_state_change(vdev_t *vd, int faulted, int degraded)
 		vdev_set_state(vd, B_FALSE, VDEV_STATE_HEALTHY, VDEV_AUX_NONE);
 }
 
+static inline void
+vdev_raidz_trim_append_rc(dkioc_free_list_t *dfl, uint64_t *num_extsp,
+    const raidz_col_t *rc)
+{
+	uint64_t num_exts = *num_extsp;
+	ASSERT(rc->rc_size != 0);
+
+	if (dfl->dfl_num_exts > 0 &&
+	    dfl->dfl_exts[num_exts - 1].dfle_start +
+	    dfl->dfl_exts[num_exts - 1].dfle_length == rc->rc_offset) {
+		dfl->dfl_exts[num_exts - 1].dfle_length += rc->rc_size;
+	} else {
+		dfl->dfl_exts[num_exts].dfle_start = rc->rc_offset;
+		dfl->dfl_exts[num_exts].dfle_length = rc->rc_size;
+		(*num_extsp)++;
+	}
+}
+
+/*
+ * Processes a trim for a raidz vdev. Because trims deal with physical
+ * addresses, we can't simply pass through our logical vdev addresses to
+ * the underlying devices. Instead, we compute a raidz map based on the
+ * logical extent addresses provided to us and construct new extent
+ * lists that then go to each component vdev.
+ */
+static void
+vdev_raidz_trim(vdev_t *vd, zio_t *pio, dkioc_free_list_t *dfl,
+    boolean_t auto_trim)
+{
+	dkioc_free_list_t **sub_dfls;
+	uint64_t *sub_dfls_num_exts;
+	zio_t *zio;
+
+	sub_dfls = kmem_zalloc(sizeof (*sub_dfls) * vd->vdev_children,
+	    KM_SLEEP);
+	sub_dfls_num_exts = kmem_zalloc(sizeof (uint64_t) * vd->vdev_children,
+	    KM_SLEEP);
+	zio = kmem_zalloc(sizeof (*zio), KM_SLEEP);
+	for (int i = 0; i < vd->vdev_children; i++) {
+		/*
+		 * We might over-allocate here, because the sub-lists can never
+		 * be longer than the parent list, but they can be shorter.
+		 * The underlying driver will discard zero-length extents.
+		 */
+		sub_dfls[i] = dfl_alloc(dfl->dfl_num_exts, KM_SLEEP);
+		sub_dfls[i]->dfl_num_exts = dfl->dfl_num_exts;
+		sub_dfls[i]->dfl_flags = dfl->dfl_flags;
+		sub_dfls[i]->dfl_offset = dfl->dfl_offset;
+		/* don't copy the check func, because it isn't raidz-aware */
+	}
+
+	/*
+	 * Process all extents and redistribute them to the component vdevs
+	 * according to a computed raidz map geometry.
+	 */
+	for (int i = 0; i < dfl->dfl_num_exts; i++) {
+		uint64_t start = dfl->dfl_exts[i].dfle_start;
+		uint64_t length = dfl->dfl_exts[i].dfle_length;
+		uint64_t j;
+		raidz_map_t *rm;
+
+		zio->io_offset = start;
+		zio->io_size = vdev_raidz_psize_floor(vd, length);
```

Comment:
Good point!  This is the same as the case in `vdev_raidz_io_start()` where optional IOs are generated for skip sectors to to improve the aggregation.  Since we have a minimum TRIM size we definitely don't way  to leave any gaps.  @thegreatgazoo @skiselkov here's an initial proposed fix for comment, I'll test it locally before updating the PR.

```diff
@@ -2402,7 +2403,7 @@ vdev_raidz_trim(vdev_t *vd, zio_t *pio, dkioc_free_list_t 
        for (int i = 0; i < dfl->dfl_num_exts; i++) {
                uint64_t start = dfl->dfl_exts[i].dfle_start;
                uint64_t length = dfl->dfl_exts[i].dfle_length;
-               uint64_t j;
+               uint64_t c, j, k;
                raidz_map_t *rm;
 
                zio->io_offset = start;

@@ -2417,6 +2418,20 @@ vdev_raidz_trim(vdev_t *vd, zio_t *pio, dkioc_free_list_t
                        vdev_raidz_trim_append_rc(sub_dfls[devidx],
                            &sub_dfls_num_exts[devidx], &rm->rm_col[j]);
                }
+
+               /*
+                * Make sure to include any skipped sectors to improve
+                * aggregation contiguity.
+                */
+               for (c = rm->rm_skipstart, k = 0; k < rm->rm_nskip; c++, k++) {
+                       ASSERT(c <= rm->rm_scols);
+                       if (c == rm->rm_scols)
+                               c = 0;
+                       uint64_t devidx = rm->rm_col[c].rc_devidx;
+                       vdev_raidz_trim_append_rc(sub_dfls[devidx],
+                           &sub_dfls_num_exts[devidx], &rm->rm_col[c]);
+               }
+
                vdev_raidz_map_free(rm);
        }

```